### DISCUSSION TITLE
Correct build process from source

### DISCUSSION THREAD

**Discussion by jimbro1000** [JanssenProject/jans#5664] [Q&A] [answered]
Transferring conversation from Gitter:

I need to be able to build jans from source (as a requirement from my SRA team)
Attempting this just using Maven results in build errors because the test suite does not differentiate unit tests and integration tests so everything is running under the surefire plugin
The build process also references external services for the purposes of testing (such as token validation), spinning up a server internally for this purpose would be acceptable using docker-in-docker but be default those external services would not be available to the build runner
Some guidance on how to correctly build from source, to jar is needed (I've yet to try the dockerfiles but for now I'm assuming that once I have jar files they are adaptable to suit the hardened base images I need to use)

**Comment by ossdhaval** [comment, marked as answer]
Hi @jimbro1000 , Welcome to Janssen Project discussions. And thanks for moving our discussion here from Gitter. 

Running builds with tests may involve few steps for some of the modules. These steps are called `Profile setup`. Steps essentially provide the details of the remote Janssen server to the workspace so that tests can be run. 
Here is a list of modules that require/not require profile setup:

No profile setup needed:
- jans-core
- jans-link
- jans-orm 
- agama


Needs profile setup:
- jans-auth-server ([Steps to setup the profile](https://docs.jans.io/head/contribute/development/#run-integration-tests-with-a-janssen-server-vm))
- jans-config-api (WIP for documenting profile setup steps)
- jans-scim (WIP for documenting profile setup steps)
- jans-fido (WIP for documenting profile setup steps)
- jans-eleven (WIP for documenting profile setup steps)

For docker related question, I'll tag @moabu .

**Reply by moabu** [reply]
@jimbro1000 Once you have the builds posted there is an env you can place inside the dockerfiles to switch to your maven repo builds. 

Generally , it's a simple make command inside each of the docker folders and it's supported already

**Reply by jimbro1000** [reply]
The guide on preparing certificates seems to be broken (I tried this before raising the question in the first place). The first step to extract the certificate from test.local.jans.io results in a zero byte length file because the domain is not valid

**Reply by jimbro1000** [reply]
I'm also very confused as to why you have your unit tests and integration tests coupled in the same tests. Surely it would be better to use the existing maven convention of putting the integration tests under failsafe instead of surefire so they run into two different lifecycle steps (and can be run independently through maven profiles in a pipeline). That would also give you control over an ephemeral environment for integration testing with appropriate configuration instead of manually setting up "profiles" for each component

**Reply by ossdhaval** [reply]
@jimbro1000 Thank you for your suggestion about separating unit and integ tests. And agreed. :handshake: 

It was always part of core team's internal discussions. And as with any refactoring, we thought it is better to chip away at this instead of a big-bang refactor. Hence we never formally tracked this effort. Now we have created an [issue](https://github.com/JanssenProject/jans/issues/6737) to do so. 

Again, thank you for your suggestion. And you are welcome to contribute as well :smiley: 

cc: @yuriyz
